karma.conf.js

```diff
@@ -0,0 +1,143 @@
+  Android11_ChromeLatest: { device: 'Google Pixel 4', os: 'Android', os_version: '11.0', browser: 'Chrome', browser_version: 'latest-beta' },
+  iOS10_Safari: { device: 'iPhone 7', os: 'iOS', os_version: '10', browser: 'Safari' },
+  iOS11_Safari: { device: 'iPhone 8 Plus', os: 'iOS', os_version: '11', browser: 'Safari' },
+  iOS12_Safari: { device: 'iPhone XS', os: 'iOS', os_version: '12', browser: 'Safari' },
+  iOS13_Safari: { device: 'iPhone 11 Pro', os: 'iOS', os_version: '13', browser: 'Safari' },
+  iOS14_Safari: { device: 'iPhone 11', os: 'iOS', os_version: '14', browser: 'Safari' },
+  iOS15_Safari: { device: 'iPhone 11 Pro', os: 'iOS', os_version: '15', browser: 'Safari' }
+}
+/* eslint-enable max-len */
+
+function makeBuildNumber() {
+  return `No CI ${Math.floor(Math.random() * 1e10)}`
+}
+
+function setupLocal(config) {
+  config.set({
+    frameworks: ['jasmine', 'karma-typescript'],
+    files: [
+      // The polyfills are required for old supported browsers.
+      // They should be removed when the old browser support is dropped.
+      'node_modules/promise-polyfill/dist/polyfill.js',
+
+      'src/**/*.ts',
+      'tests/**/*.ts',
+      'test-dist/botd.min.js',
+    ],
+    preprocessors: {
+      '**/*.ts': 'karma-typescript',
+    },
+    reporters: ['spec', 'summary'],
+    browsers: ['ChromeHeadless', 'FirefoxHeadless'],
+    concurrency: 3,
+
+    karmaTypescriptConfig: {
+      compilerOptions: {
+        ...require('./tsconfig.json').compilerOptions,
+        module: 'commonjs',
+        sourceMap: true,
+      },
+    },
+
+    specReporter: {
+      suppressSummary: true,
+      suppressErrorSummary: true,
+      suppressPassed: true,
+      suppressSkipped: true,
+    },
+
+    summaryReporter: {
+      show: 'skipped', // To know that some tests are skipped exactly where they are supposed to be skipped
+    },
+  })
+}
+
+function setupBrowserstack(config) {
+  setupLocal(config)
+
+  const customLaunchers = {}
+  for (const [key, data] of Object.entries(browserstackBrowsers)) {
+    customLaunchers[key] = {
+      base: 'BrowserStack',
+      name: key.replace(/_/g, ' '),
+      ...data,
+    }
+  }
+
+  config.set({
+    reporters: [...config.reporters, 'BrowserStack'],
+    browsers: Object.keys(customLaunchers),
+    customLaunchers,
+    concurrency: 5,
+    failOnEmptyTestSuite: false,
+
+    browserStack: {
+      project: 'BotD',
+      // A build number is required to group testing sessions in the BrowserStack UI.
+      // GitHub Actions will add a value for GITHUB_RUN_ID. More on the environment variables:
+      // https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables#default-environment-variables
+      build: process.env.GITHUB_RUN_ID || makeBuildNumber(),
+      // The timeout is reduced for testing sessions to not hold the BrowserStack queue long in case of problems.
+      timeout: 5 * 60,
```

Comment:
Should be as little as necessary. When BrowserStack is stuck (it happens sometimes), everybody will have to wait 5 minutes until the case times out and the browser queue is released before they can run their tests.